applications/yatima.md

```diff
@@ -0,0 +1,439 @@
+  and that they are not just replacing counterparty risk with technical risk. We
+  believe that functional programming and formal proofs offer a possible path
+  towards increasing the safety and reliability of all software, not just in
+  blockchain.
+
+### Project Details
+
+Many blockchain projects are exploring strongly-typed functional programming as
+a paradigm for writing smart contracts and DApps. Examples include [Cardano's
+Plutus](https://developers.cardano.org/en/programming-languages/plutus/overview/), [Tezos' Michelson](https://wiki.tezosagora.org/learn/smartcontracts/michelson), [Kadena's Pact](https://github.com/kadena-io/pact) and [Digital Asset's DAML](https://daml.com/). The type safety and
+composability of functional programs make them an exciting option for
+writing high-assurance smart contracts.
+
+Yatima aims to help develop the functional programming ecosystem on Substrate
+and the Polkadot ecosystem. We have developed a graph-based runtime for the Yatima language
+with fully lazy evaluation, sharing, and whole program optimization. This runtime could also 
+be targeted by other functional languages. The Yatima runtime is
+similar in many respects to Haskell Core, or the [GRIN
+IR](https://grin-compiler.github.io/) but is implemented in deterministic
+`no_std` Rust, which allows it to run virtually anywhere, including the web,
+WASM blockchains, or even embedded systems (in theory).
+
+Yatima also implements a powerful dependent and substructural type-system,
+similar to Idris 2. This enables formal verification and theorem proving
+(although crucial features for practical proving such as inference and tactics
+are not yet implemented). Proofs allow for many useful optimizations in the area
+of smart contracts, such as static detection of arithmetic overflow or division
+by zero errors, or tracking of resources and permissions via an effect system.
+Furthermore, since Yatima can be implemented in itself, there is a possiblity of
+having on-chain trustless proof verification.
+
+Lastly, Yatima implements content-addressing, allowing all programs and packages
+to be serialized into IPLD objects and shared over the peer-to-peer IPFS network
+(or via libp2p).
+
+#### Deliverables
+
+- A `sp_ipld` crate for interacting with IPFS CID, Multihash and IPLD based on
+  [`ipfs-rust/libipld`](https://github.com/ipfs-rust/libipld), but which is
+  Substrate pallet compatible. This is a necessary dependency for Yatima Core,
+  but is an independent deliverable that could be useful for many other
+  Substrate projects.
+
+- A `sp_im` crate for creating immutable datastructures, based on
+  [`bodil/im-rs`](https://github.com/bodil/im-rs), but which is
+  Substrate pallet compatible. This is a necessary dependency for Yatima Core,
+  but is an independent deliverable that could be useful for many other
+  Substrate projects.
+
+- A `sp_yatima` crate integrated with `sp_ipld`, `sp_im` and `sp_std` that
```

Comment:
for demonstration purposes, could you include rudimentary integration with `sp-storage`? Just read-level operations. I think that would be useful for a follow-up grant motivation.

Reply:
I agree that integrating with `sp-storage` would be useful, but integrating read operations directly into the language is not actually much less work than integrating both read and write operations. In both cases we're effectfully interacting with an external DB, and have to correctly compute key values and correctly model StorageValue return types, both of which seem non-trivial. We have been considering two different approaches to this, which may not be mutually exclusive:

1. Creating a custom SubstrateIO type in Yatima which mirrors the APIs found in https://substrate.dev/rustdocs/v3.0.0/frame_support/storage/trait.StorageMap.html#required-methods and https://substrate.dev/rustdocs/v3.0.0/frame_support/storage/trait.StorageValue.html.
2. Extend yatima_core with first-class-modules primitives (dependently typed maps) and use the modules as a kind of frontend to the StorageMaps without requiring everything to be mapped to monadic IO.

Both cases involve extending our runtime DAG machine with primitives that actually perform the right Substrate calls, so that's probably where we'll start. Not 100% sure it's really in scope as part of the current proposal, which is essentially "get the pure language working", but it's definitely Milestone 1 on our follow-up.

Reply:
Thanks for the clarification. On a high-level, I'd prefer the second approach, but let's cross that bridge once we get there.